File: sdk/java/gladys-sensor-sdk/src/test/java/com/gladys/sensor/TestSensorHarnessTest.java
package com.gladys.sensor;

import com.gladys.sensor.testing.TestSensorHarness;
import gladys.v1.Common;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class TestSensorHarnessTest {

    private TestSensorHarness harness;

    @BeforeEach
    void setUp() {
        CommandDispatcher dispatcher = CommandDispatcher.builder()
                .onStart(args -> null)
                .onStop(args -> null)
                .onPause(() -> null)
                .onResume(() -> null)
                .onReload(() -> null)
                .onHealthCheck(args -> null)
                .onRecover(args -> null)
                .build();

        harness = new TestSensorHarness(dispatcher);
    }

    @Test
    void testDispatchAllSevenCommands() {
        // START -> ACTIVE
        CommandDispatcher.DispatchResult result = harness.dispatchStart(StartArgs.testArgs(false));
        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, harness.getState());

        // PAUSE -> PAUSED
        result = harness.dispatchPause();
        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_PAUSED, harness.getState());

        // RESUME -> ACTIVE
        result = harness.dispatchResume();
        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, harness.getState());

        // RELOAD -> ACTIVE
        result = harness.dispatchReload();
        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, harness.getState());

        // HEALTH_CHECK -> unchanged (ACTIVE)
        result = harness.dispatchHealthCheck(HealthCheckArgs.testArgs(false));
        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, harness.getState());

        // RECOVER -> ACTIVE
        result = harness.dispatchRecover(RecoverArgs.testArgs("default"));
        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, harness.getState());

        // STOP -> STOPPED
        result = harness.dispatchStop(StopArgs.testArgs(false, false));
        assertFalse(result.hasError());
        assertEquals(Common.ComponentState.COMPONENT_STATE_STOPPED, harness.getState());
    }

    @Test
    void testGetSetState() {
        // Initial state is UNKNOWN
        assertEquals(Common.ComponentState.COMPONENT_STATE_UNKNOWN, harness.getState());

        // Set state directly
        harness.setState(Common.ComponentState.COMPONENT_STATE_ACTIVE);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ACTIVE, harness.getState());

        // Set to ERROR
        harness.setState(Common.ComponentState.COMPONENT_STATE_ERROR);
        assertEquals(Common.ComponentState.COMPONENT_STATE_ERROR, harness.getState());
    }
}

File: sdk/js/gladys-sensor-sdk/test/TestSensorHarness.test.ts
import { describe, it, expect, beforeEach } from "vitest";
import { TestSensorHarness } from "../src/testing/TestSensorHarness";
import { CommandDispatcher } from "../src/CommandDispatcher";
import { ComponentState } from "../src/generated/common";

describe("TestSensorHarness", () => {
  let dispatcher: CommandDispatcher;
  let harness: TestSensorHarness;

  beforeEach(() => {
    dispatcher = new CommandDispatcher()
      .onStart(async (args) => {
        if (args.dryRun) return ComponentState.COMPONENT_STATE_STARTING;
        return null;
      })
      .onStop(async () => null)
      .onPause(async () => null)
      .onResume(async () => null)
      .onReload(async () => null)
      .onHealthCheck(async () => null)
      .onRecover(async () => null);

    harness = new TestSensorHarness(dispatcher);
  });

  it("dispatches START", async () => {
    const result = await harness.dispatchStart();

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
  });

  it("dispatches STOP", async () => {
    harness.setState(ComponentState.COMPONENT_STATE_ACTIVE);

    const result = await harness.dispatchStop();

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_STOPPED);
  });

  it("dispatches PAUSE", async () => {
    harness.setState(ComponentState.COMPONENT_STATE_ACTIVE);

    const result = await harness.dispatchPause();

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_PAUSED);
  });

  it("dispatches RESUME", async () => {
    harness.setState(ComponentState.COMPONENT_STATE_PAUSED);

    const result = await harness.dispatchResume();

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
  });

  it("dispatches RELOAD", async () => {
    harness.setState(ComponentState.COMPONENT_STATE_ACTIVE);

    const result = await harness.dispatchReload();

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
  });

  it("dispatches HEALTH_CHECK", async () => {
    harness.setState(ComponentState.COMPONENT_STATE_ACTIVE);

    const result = await harness.dispatchHealthCheck();

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
  });

  it("dispatches RECOVER", async () => {
    harness.setState(ComponentState.COMPONENT_STATE_ERROR);

    const result = await harness.dispatchRecover();

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
  });

  it("get and set state", () => {
    expect(harness.getState()).toBe(ComponentState.COMPONENT_STATE_UNKNOWN);

    harness.setState(ComponentState.COMPONENT_STATE_ACTIVE);

    expect(harness.getState()).toBe(ComponentState.COMPONENT_STATE_ACTIVE);
  });

  it("default args when none provided", async () => {
    // START with no args should use default dryRun=false, resulting in ACTIVE
    const result = await harness.dispatchStart();

    expect(result.state).toBe(ComponentState.COMPONENT_STATE_ACTIVE);

    // START with dryRun=true should return STARTING (handler override)
    harness.setState(ComponentState.COMPONENT_STATE_UNKNOWN);
    const result2 = await harness.dispatchStart({ dryRun: true });

    expect(result2.state).toBe(ComponentState.COMPONENT_STATE_STARTING);
  });
});
